src/Symfony/Component/Message/DependencyInjection/MessagePass.php

```diff
@@ -79,7 +81,22 @@ private function findHandlers(ContainerBuilder $container): array
             $handlersByMessage[$message] = call_user_func_array('array_merge', $handlersByMessage[$message]);
         }
 
-        return $handlersByMessage;
+
+        $definitions = array();
+        foreach ($handlersByMessage as $message => $handlers) {
+            if (1 === count($handlers)) {
+                $handlersByMessage[$message] = current($handlers);
+            } else {
+                $d = new Definition(MessageHandlerCollection::class, array($handlers));
+                $d->setPrivate(true);
+                $definitions[hash('sha1', $message)] = $d;
```

Comment:
Do we need to prefix the service name or are we sure that sha1 is unique enough?
If we go for prefix. Suggestions. I'd add `MHC` short for `MessageHandlerCollection`

Reply:
Why not `MessageHandlerCollection` ? :-)

Reply:
The prefix? I find it rather long 40 characters of sha1 + 24 for prefix. We'll have promlems on windows when this gets dumped to filesystem with the new container (aka service per file)

Reply:
Fair enough, let's start with that and then we'll see the reviews in Symfony's main PR anyway :)